lebedigital/calibration/utils.py

```diff
@@ -0,0 +1,161 @@
+
+    Parameters
+    ----------
+    path : str Path to the folder where the experimental data in .csv format is stored.
+    exp_name : str the experiment name.csv
+    length : float the length of the specimen
+    diameter : the diameter of the specimen
+
+    Returns
+    -------
+    output : dict
+    values with the keys : exp_name, height, diameter, displacement, force
+    """
+    file_path = os.path.join(path, exp_name)
+    results = {}
+    results["length"] = length
+    results["diameter"] = diameter
+    df = extract_third_load_cycle(data_path=file_path)
+    df["displacement"] = (
+        df["Transducer 1[mm]"] + df["Transducer 2[mm]"] + df["Transducer 3[mm]"]
+    ) / 3
+    # df['stress'] = df['Force [kN]'] / (np.pi * (float(results['diameter']) / 2) ** 2)
+
+    output = {
+        "exp_name": exp_name,
+        "height": float(results["length"]),
+        "diameter": float(results["diameter"]),
+        "displacement": np.array(df["displacement"]),
+        "force": np.array(df["Force [kN]"]),
+    }
+    return output
+
+
+class PosteriorPredictive:
+    def __init__(
+        self, forward_solver, known_input_solver, parameter=None
+    ):
+
+        """
+
+        Parameters
+        ----------
+        forward_solver :
+            Should be wrapped in a function which takes in two args. arg1 = random input, arg2 = known input
+        parameter :
+            An array of samples. Currently parameters dont support a distrutbion, just MCMC samples. To get an approx.
+            for parameter dist, we need VI based posterior approx
+        known_input_solver :
+            Pass known input value to the solver in whichever format the forward_solve object needs.
+        """
+
+        # assert (parameter == np.array), "Parameter needs to be np.ndarray"
+        # assert parameter.shape[1] == 1, "Currently support only 1 parameter"
+
+        self._forward_solver = forward_solver
+        if parameter is not None:
+            self._parameter = parameter
+        self._known_input = known_input_solver
+        self._mean = None
+        self._std = None
+        self._samples = None
+
+    def get_stats(self, samples: int) -> tuple:
+        """
+        Returns mean and s.d of the posterior predictive. Simple Monte Carlo based approximation
+
+        Parameters
+        ----------
+        samples :
+
+        Returns
+        -------
+
+        """
+        # Monte carlo step
+        output = []
+        for i in range(0, samples):
+            for i in range(0, samples):
+                y = self._forward_solver(self._parameter[i], self._known_input)
+                output.append(y)
+
+        # get the posterior pred stats
+        mean = np.mean(output, axis=0)
+        sd = np.std(output, axis=0)
+
+        self._mean = mean
+        self._std = sd
+        self._samples = output
+        return mean, sd
+
+    def plot_posterior_predictive(self):
+        raise NotImplementedError("...")
```

Comment:
for later day, if I include the plot method here.

Reply:
thats fine, maybe add a comment so that its clear.
although I guess we would rather use the data stored in the KG to then generate a plot as a seperate step. but we dont need to worry about that now.

Reply:
we already discussed this over chat